README.md
# GeoIPMap

This program plots the geographical coordinates of IPs (v4) using a
database (CSV) in a world map using the Mercator projection.

  * **Version:** 1.2.0 (2019-03-24)
  * **Requires:** [`matplotlib`](https://matplotlib.org/),
                  [`netaddr`](https://pypi.org/project/netaddr).


## Installing dependencies

Many Linux distributions already have a package manager that allows to
download this dependencies.

  * Debian:

        $ sudo apt-get install python-matplotlib
        $ sudo apt-get install python-netaddr

  * Gentoo:

        $ sudo emerge -a matplotlib
        $ sudo emerge -a netaddr

  * Arch Linux:

        $ sudo pacman -S python-matplotlib
        $ sudo pacman -S install python-netaddr

If that's not the case, you can always install them by using the `pip`
tool.

        $ python -m pip install -U matplotlib
        $ python -m pip install -U netaddr


## Basics

All needed data could be conveniently contained in the folder `data/`,
although through arguments you can specify another files or path.  The
usage of the main entry is:

        -i <IPLIST>, --iplist <IPLIST>
                            File of IPs to plot, one per line
        -g <GEODB>, --geodb <GEODB>
                            CSV document with columns: 'network,lat,lon'
        -f <IMAGEFILE>, --imagefile <IMAGEFILE>
                            Path for the template image to plot over
        -W <WIDTH>, --width <WIDTH>
                            Width in pixels of the image file
        -H <HEIGHT>, --height <HEIGHT>
                            Height in pixels of the image file
        -l <LEFT>, --left <LEFT>
                            Degrees most to the left in the image
        -r <RIGHT>, --right <RIGHT>
                            Degrees most to the right in the image
        -b <BOTTOM>, --bottom <BOTTOM>
                            Degrees most to the bottom in the image
        -n <NSPLITS>, --nsplits <NSPLITS>
                            Splits for reading 'geodb' file (threading)

You can see this options anytime by using the `-h` or `--help` argument.
The default values for the options `-l`, `-r`, and `-b` are calibrated
for the Mercator map with resolution of 2058×1746 pixels you can find in
Wikipedia, as explained below.

Three files are required for this program to work:

  * **Image file.**  This is a world map image in the Mercator
    projection.  The file properties such as `width`, `length`, and the
    degrees of left, right, and bottom boundaries depicted may be
    entered through the command line, as well as the filename.  When
    changing the image, you have to adjust three points specified in
    degrees, the point most to the left (by default is -180°), the point
    most to the right (by default is 180°), and the point most to the
    south.  It shouldn't be greater than -90°, although if you have a
    map where the most fore south point is exactly at -90°, you should
    use an approximation, such as -89.99, or a division by zero may
    happen because the linear scale becomes infinitely large at the
    poles in this projection.

    The default map for the configured settings is the one you can find
    in Wikipedia for the article: "[Mercator
    Projection](https://en.wikipedia.org/wiki/Mercator_projection)", but
    in the size of 2058×1746 pixels.  It's set between 82°S and 82°N,
    *i.e.*, with latitudes in the range [-82, 82].

    You can download and use a different map, or the same map in a
    different resolution, but you'll have to specify the dimensions and
    other parameters through the command line when invoking the program.

  * **IP list.**  A file of the IPs you want to plot in a map.  There
    have to be IP addresses, not in subnet notation or netmasks, one
    per line.  By default, the file used will be `data/ips.lst`.  For
    example:

        ...
        77.247.181.162
        198.98.56.149
        185.220.100.252
        162.247.74.202
        195.176.3.19
        145.239.91.37
        185.220.101.54
        51.15.125.181
        ...

  * **Geocoordinates database.**  A CSV file containing the following
    columns, without comments (may be this will be solved in a future
    version): `netmask, latitude, longitude`.  The default CSV file used
    is `data/geoip_ipv4.csv`.  For example:

        ...
        73.185.104.0/23,37.8209,-121.2827
        73.185.106.0/23,38.6106,-121.2789
        73.185.108.0/23,38.5879,-121.4053
        73.185.110.0/23,38.6711,-121.1495
        73.185.112.0/23,36.3170,-119.3087
        73.185.114.0/23,38.1345,-120.4516
        73.185.116.0/23,39.1663,-121.5105
        ...

## Quick usage

  1. Download the Mercator map (2058×1746 pixels) from Wikipedia:
    <https://upload.wikimedia.org/wikipedia/commons/f/f4/Mercator_projection_SW.jpg>
    and save it in the `data` folder under the name of
    `worldmap_raw.jpg`.

  2. Put a file containing the IPs you want to plot (one per line) in
     the map in the folder `data` under the name of `ips.lst`.

  3. Create a CSV database with the columns:
     `netmask,latitude,longitude` and save it in the folder `data`under
     the name `geoip_ipv4.csv`.  See below.


## Generating a CSV database

The CSV database of 'network, longitude, latitude' is not included due
its weight, but it can be constructed easily.

  1. Download from [MaxMind](https://dev.maxmind.com) the GeoLite2 City
     database in CSV format.  You'll find it at:
  <https://geolite.maxmind.com/download/geoip/database/GeoLite2-City-CSV.zip>.

  2. Uncompress that file and work with the
     `GeoLite2-City-Blocks-IPv4.csv` file.  It's a very huge file and
     has much more information that the needed for this program to run.

  3. Create a new document using the following command:

        `$ awk -F',' '{print $1","$8","$9}' GeoLite2-City-Blocks-IPv4.csv >geoip_ipv4.csv`

     Put the output file in the `data` directory with the name
     `geoip_ipv4.csv`.

## Using custom files

The only requirement is the format of the files and the projection used
in the map.  You can use different filenames specifying them in the
option arguments:

Instead `python main.py`, you may use:

        $ python main.py -i <my_iplist> -g <my_geodb> -f <my_image> -b -80

When changing the image, you have to specify, at least, the degrees more
to the south the map displays (and in some cases, the points more to the
west and/or to the right when the map is not calibrated in between 180°W
and 180°E, *i.e.*, the range [-180, 180], for degrees of longitude.


## Output and display

This is an example of the output provided by this program using
`matplotlib`.  This image is a zoomed detail.

<!-- ![Output example](images/output_example.png) -->
![Output detail example](images/output_example_detail.png
    "Output detail example")

The red dots indicate each one of the locations of the IPs in the file
`ips.lst`.  The radius of this circle is calculated automatically in the
function `plot_pixels` (in the `geoipmap.py` file).

The automatic calculation provides a way to draw the circles with a size
in relation to the image size.  The computation is performed by getting
the biggest dimension of the image, and set the radius to a 0.3% of this
length, so, for the most common resolutions:

| Resolution  | Radius |
|:-----------:|-------:|
|   800×600   |  2.400 |
|  1024×768   |  3.072 |
|  1280×1024  |  3.840 |
|  1920×1080  |  5.760 |
|  3840×2160  | 11.520 |
|  7680×4320  | 23.040 |

The default image detailed above has a resolution of 2058×1746 pixels,
so its radius will be 6.174 pixels.

This default behaviour can be altered by invoking the `plot_pixels`
function with an extra argument indicating the float value of the radius
in pixels.

        # Automatic radius
        geo.plot()

        # User defined radius
        geo.plot(radius=10)

This function is invoked in the main entry, at the end of the file
`main.py`.

**NOTE.**  This method will be probably changed in the future to a much
better approximation.


## Multiprocessing and threading

See the [benchmarks](benchmark.md) document.


## License

This program is under the BSD 2-Clause license.  This license has also
been called the "Simplified BSD License" and the "FreeBSD License".

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

      1. Redistributions of source code must retain the above copyright
         notice, this list of conditions and the following disclaimer.

      2. Redistributions in binary form must reproduce the above
         copyright notice, this list of conditions and the following
         disclaimer in the documentation and/or other materials provided
         with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.

The BSD 2-clause license allows you almost unlimited freedom with the
software so long as you include the BSD copyright notice in it (found in
the `LICENSE` file).

Under this license, you may:

  * use the software for commercial use;
  * modify the software and create derivatives;
  * distribute original or modified works;
  * place warranty on the software licensed.

You must:

  * include copyright
  * include the full text of license in modified software.

More information about the license:
<https://en.wikipedia.org/wiki/BSD_licenses#2-clause>.


---

J. A. Corbal, 2019.



CHANGELOG
Changelog
=========

v 1.2.0 (2019-03-24):
  * Added multiprocessing support

v 1.1.2-rc.1 (2019-03-23):
  * GeoIpMap is now a class and not separated functions
  * Threading support (not very efficient)
  * Bugs fixed

v 1.0.1 (2019-03-21):
  * Added matplotlib support



geoipmap/__init__.py
#!/usr/bin/env python
# vim: set ft=python fenc=utf-8 tw=72:

# GEOIPMAP :: World map plotting of locations associated to IPs
# Copyright (c) 2019, J. A. Corbal
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the
#    distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""
    GeopIPMap
    ~~~~~~~~~

    Plots in a world map the geographical locations of IPs

    :author: J. A. Corbal
    :copyright: (C) 2019, J. A. Corbal
    :license: BSD 2-Clause "Simplified" License
"""

from geoipmap.geoimage import GeoImage
from geoipmap.geoipmap import GeoIpMap


geoipmap/geoimage.py
#!/usr/bin/env python
# vim: set ft=python fenc=utf8 fo=tcqrj1n:

# GEOIPMAP :: World map plotting of locations associated to IPs
# Copyright (C) 2019, J. A. Corbal <jacorbal@gmail.com>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#   1. Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#   2. Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in
#      the documentation and/or other materials provided with the
#      distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""Just a image class, more like a container for image information"""

__author__ = "J. A. Corbal"
__copyright__ = "Copyright (C) 2019, J. A. Corbal"
__license__ = "BSD 2-Clause"
__version__ = "1.2.0"
__mantainer__ = "J. A. Corbal"
__email__ = "jacorbal@gmail.com"
__status__ = "Development"


class GeoImage(object):
    """Image class (more like a struct)."""
    def __init__(self, filepath, width, height,
                 w_deg=-180, e_deg=180, s_deg=-89.9):
        """Geo image data:

        :param filepath: Path to the image
        :type filepath: str
        :param width: Width in pixels
        :type width: int
        :param height: Height in pixels
        :type height: int
        :param w_deg: Most fore west position in degrees
        :type w_deg: float
        :param e_deg: Most fore east position in degrees
        :type e_deg: float
        :param s_deg: Most fore south position in degrees
        :type s_deg: float
        """
        self.__set_filepath(filepath)
        self.__set_width(width)
        self.__set_height(height)
        self.__set_w_deg(w_deg)
        self.__set_e_deg(e_deg)
        self.__set_s_deg(s_deg)

    def __get_filepath(self):
        return self._filepath

    def __set_filepath(self, filepath):
        if filepath:
            self._filepath = filepath

    def __get_width(self):
        return self._width

    def __get_height(self):
        return self._height

    def __get_w_deg(self):
        return self._w_deg

    def __get_e_deg(self):
        return self._e_deg

    def __get_s_deg(self):
        return self._s_deg

    def __set_width(self, width):
        if (width > 0):
            self._width = width

    def __set_height(self, height):
        if (height > 0):
            self._height = height

    def __set_w_deg(self, w_deg):
        self._w_deg = w_deg

    def __set_e_deg(self, e_deg):
        self._e_deg = e_deg

    def __set_s_deg(self, s_deg):
        self._s_deg = s_deg

    # Properties
    filepath = property(__get_filepath, __set_filepath)
    width = property(__get_width, __set_width)
    height = property(__get_height, __set_height)
    w_deg = property(__get_w_deg, __set_w_deg)
    e_deg = property(__get_e_deg, __set_e_deg)
    s_deg = property(__get_s_deg, __set_s_deg)

    # Class printing
    def __repr__(self):
        return self._filepath

    def __str__(self):
        return "'{}' ({}x{})".format(self._filepath, self._width, self._height)


geoipmap/geoipmap.py
#!/usr/bin/env python
# vim: set ft=python fenc=utf8 fo=tcqrj1n:

# GEOIPMAP :: World map plotting of locations associated to IPs
# Copyright (C) 2019, J. A. Corbal <jacorbal@gmail.com>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#   1. Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#   2. Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in
#      the documentation and/or other materials provided with the
#      distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""
Methods to gather longitude and latitude coordinates given an IP address
using a database (CSV) and plot those points into a world map using the
Mercator projection.

Required:
  * List of IPs (one IP per line)
  * Geolocalization database in the format of a CSV with the following
    columns: 'network,latitude,longitude', where latitude and longitude
    are floating point degrees.
  * Map image.  If the properties change, those changes must be
    addressed in this file.
"""

__author__ = "J. A. Corbal"
__copyright__ = "Copyright (C) 2019, J. A. Corbal"
__license__ = "BSD 2-Clause"
__version__ = "1.2.0"
__mantainer__ = "J. A. Corbal"
__email__ = "jacorbal@gmail.com"
__status__ = "Development"


import csv
import math
import matplotlib.patches
import matplotlib.pyplot as plt
import multiprocessing
from netaddr import IPAddress
from netaddr import IPNetwork
from netaddr import valid_ipv4
import threading


class GeoIpMap(object):
    """GeoIpMap class.  Get the data, calculate and plot.

    This class instantiates objects capable of the following methods:

      1. Get list of IPs and database of IPs and geocoordinates
      2. Convert IP addresses to pixels: `ips_to_pixels`
      3. Plot the pixels in a image: `plot`

    In order to perform this actions, three files are needed:

      * `ips_file`.  A list of the IPs to plot, one per line
      * `geodb_file`.  A CSV file with columns: 'netmask,lat,lon'
      * Image file.  Contained indirectly in the `img` object

    ..note:: Right now, the files must be formatted with no comments.

    This class returns the `_pixels` set attribute when invoked for
    printing, containing the pairs (`x`, `y`) to be drawn in the map.
    """

    def __init__(self, ips_file, geodb_file, img,
                 method='multiprocessing',  num_workers=4,
                 num_splits=10, verbose=False):
        """Initializes the GeoIPMap object.

        If the method is 'threads', then the method for finding the IPs
        in the table with be done by the `threading` module (by the
        applying `__ips_to_pixels_threading`); otherwise it will use the
        `multiprocessing` module (using `__ips_to_pixels_multiprocessing`).
        If `num_splits` is less or equal to one, none of those methods
        will be used, and the table search is done in the "classical
        way" (with `__ips_to_pixels_raw`).

        When using `method=threading`, the argument `num_splits`
        indicate in how many chunks the list is divided in order to make
        better comparisons.  This variable is nonfunctional when using
        the multiprocessing method.

        When using `method!=threading` (i.e., `method=multiprocessing`),
        the argument `num_workers` indicate how many workers are being
        used in the calculation.  This variable is nonfunctional when
        using the threading method.

        When using `method=raw`, then no multiprocessing or threading is
        performed.

        :param iplist_file: IP list file, one IP per line
        :type iplist_file: str
        :param geodb_file: Netmask, latitude and longitude CSV database
        :type geodb_file: str
        :param img: Image object with information to plot
        :type img: GeoImage
        :param num_splits: Splits of `geodb_file` for reading (threading)
        :type num_splits: int
        :param num_workers: Number of workers when multiprocessing
        :type num_workers: int
        :param verbose: When true, print information on screen
        :type verbose: bool
        """
        self._pixels = set()
        self._img = img
        self._method = method
        self._verbose = verbose
        self.__set_num_splits(num_splits)
        self.__set_num_workers(num_workers)
        self.__set_geodb(geodb_file)
        self.__set_ips(ips_file)

    def __get_num_splits(self):
        return self._num_splits

    def __get_num_workers(self):
        return self._num_workers

    def __get_geodb(self):
        return self._geodb

    def __get_ips(self):
        return self._ips

    def __set_geodb(self, geodb_file):
        if (self._verbose):
            print('Unpacking', geodb_file)
        with open(geodb_file, 'rt') as f:
            reader = csv.reader(f)
            self._geodb = list(reader)

    def __set_ips(self, ips_file):
        if (self._verbose):
            print('Unpacking', ips_file)
        with open(ips_file, 'rt') as f:
            self._ips = f.readlines()
            self._ips = [x.strip() for x in self._ips]

    def __set_num_splits(self, num_splits):
        """Sets the number of splits the threading function can do."""
        if num_splits < 0:
            self._num_splits = 0
#        elif num_splits > 200:
#            self._num_splits = 200
        else:
            self._num_splits = num_splits

    def __set_num_workers(self, num_workers):
        """Sets the number of workers for multiprocessing."""
        if num_workers < 0:
            self._num_workers = 0
        else:
            self._num_workers = num_workers

    def __del_ips(self):
        del self._ips

    def __del_geodb(self):
        del self._geodb

    def geo_to_pixel(self, lat, lon):
        """Converts longitude and latitude to Mercator proj. coords.

        :param lat: Latitude in degrees
        :type lat: float
        :param lon: Longitude in degrees
        :type lon: float
        :return: Pixel coordinates (`x`, `y`)
        :rtype: pair
        """
        lat_south_rad = math.radians(self._img.s_deg)
        lat_rad = math.radians(lat)
        lon_delta = (self._img.e_deg - self._img.w_deg)
        map_width = (self._img.width / lon_delta) * 360 / (2 * math.pi)
        offset_y = (map_width /
                    2 * math.log((1 + math.sin(lat_south_rad)) /
                                 (1 - math.sin(lat_south_rad))))

        x = (lon - self._img.w_deg) * (self._img.width / lon_delta)
        y = self._img.height - \
            ((map_width / 2 * math.log((1 + math.sin(lat_rad)) /
                                       (1 - math.sin(lat_rad)))) -
             offset_y)

        return (x, y)

    def __ips_to_pixels_raw(self):
        """Return a list of pairs (`x`, `y`) from a list of IPs.

        ..note:: This method is very heavy; just for debugging.
                 The original purpose of this method was to test the
                 program, but other methods are
                 `__ips_to_pixels_threading` and
                 `__ips_to_pixels_multiprocessing`.

        This method check the geoposition of the list of IP using a
        database file in the format CSV, then translate their
        (`lon`, `lat`) to the points in a plane.

        :see: self.geo_to_pixel
        """
        if self._verbose:
            print("Gathering data; this may take a while...")

        self._pixels = {self.geo_to_pixel(float(lat), float(lon))
                        for [nw, lat, lon] in self._geodb
                        for ip in self._ips
                        if valid_ipv4(ip) and
                        IPAddress(ip) in IPNetwork(nw)}

        if self._verbose:
            print("Data gathered!")

    def __match_geodb(self, start=None, end=None):
        """Private method to get the pixels for an IP.

        This method check the geoposition of the list of IP using a
        database file in the format CSV, then translate their
        (`lon`, `lat`) to the points in a plane.  But instead of
        searching in the whole database, it looks within a range defined
        by [`start`, `end`].  It `start` and/or `end` are `None`, then
        it will be considered repectively the beginning and the end of
        the list.

        :param start: Where to start in the CSV database
        :type start: int
        :param end: Where to end
        :type start: int
        """
        for nw, lat, lon in self._geodb[start:end]:
            for ip in self._ips:
                if valid_ipv4(ip) and IPAddress(ip) in IPNetwork(nw):
                    self._pixels.add(self.geo_to_pixel(float(lat),
                                                       float(lon)))

    def __ips_to_pixels_threading(self):
        """Splits the load of `__match_geodb` to run collectively.

        This is the threading version of `ips_to_pixels`

        ..see:: self.__ips_to_pixels_raw
        """
        if self._verbose:
            print("Threading; gathering data; this may take a while...")

        split_size = len(self._geodb) // self._num_splits
        threads = []
        for i in range(self._num_splits):
            start = i * split_size
            end = None if i + 1 == self._num_splits \
                else (i + 1) * split_size
            threads.append(
                threading.Thread(target=self.__match_geodb,
                                 args=(start, end)))
            threads[-1].start()

        for t in threads:
            t.join()

        if self._verbose:
            print("Data gathered!")

    def __ips_to_pixels_multiprocessing(self):
        """Splits the load of `__match_geodb` to run collectively.

        This is the multiprocessing version of `ips_to_pixels`

        ..see:: self.__ips_to_pixels_raw
        """
        if self._verbose:
            print("Multiprocessing; gathering data; this may take a while...")

        workers = []
        for i in range(self._num_workers):
            workers.append(multiprocessing.Process(target=self.__match_geodb))
            workers[-1].start()

        for worker in workers:
            worker.join()

        if self._verbose:
            print("Data gathered!")

    def ips_to_pixels(self):
        """Sets the default method for translating IPs to pixels.

        Depending on the class attribute `self._num_splits`, it chooses
        a method with threading/multiprocessing or without them.

        When the number of workers is one (or less), instead of
        multiprocessing the calculations, those are made by the *raw*
        method.  The same happens when using the threading method using
        one split of the data.
        """
        if self._num_splits <= 1 or self._num_workers <= 1 or \
           self._method == 'raw':
            self.__ips_to_pixels_raw()
        else:
            if self._method == 'threading':
                self.__ips_to_pixels_threading()
            else:
                self.__ips_to_pixels_multiprocessing()

    def plot(self, radius=None):
        """Plot coordinates from a list of pairs (`x`, `y`) as circles.

        When not set, the radius of the circle is obtained from the
        resolution of the image, where usually 'width > height' but
        'width ~= height'.  The default radius is the 0.3% of the
        largest dimension of the image, so:

                ============ ========
                 Resolution   Radius
                ============ ========
                  800×600      2.400
                 1024×768      3.072
                 1280×1024     3.840
                 1920×1080     5.760
                 3840×2160    11.520
                 7680×4320    23.040
                ============ ========

        The default image has a resolution of 2058×1746 pixels, so its
        radius will be 6.174.

        :param radius: radius in pixels of the circles
        :type radius: float
        """
        if not radius:
            radius = 0.003 * self._img.width \
                if self._img.width > self._img.height \
                else 0.003 * self._img.height

        img = matplotlib.image.imread(self._img.filepath)
        fig, ax = plt.subplots(1)
        ax.set_aspect('equal')
        ax.imshow(img)

        for x, y in self._pixels:
            circ = matplotlib.patches.Circle((x, y), radius,
                                             facecolor='red',
                                             edgecolor='black')
            ax.add_patch(circ)

        # Show the image
        plt.axis('off')
        plt.subplots_adjust(left=0, right=1, top=1, bottom=0)
        if self._verbose:
            print('Plotting')
        plt.show()

    # Properties
    num_splits = property(__get_num_splits, __set_num_splits)
    num_workers = property(__get_num_workers, __set_num_workers)
    ips = property(__get_ips, __set_ips, __del_ips)
    geodb = property(__get_geodb, __set_geodb, __del_geodb)

    # Class printing
    def __repr__(self):
        return self._pixels

    def __str__(self):
        return self._pixels


main.py
#!/usr/bin/env python
# vim: set ft=python fenc=utf8 fo=tcqrj1n:

# GEOIPMAP :: World map plotting of locations associated to IPs
# Copyright (C) 2019, J. A. Corbal <jacorbal@gmail.com>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#   1. Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#   2. Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in
#      the documentation and/or other materials provided with the
#      distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import argparse
from geoipmap import GeoImage
from geoipmap import GeoIpMap
import sys


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Plot coordinates of IP \
                                                  addresses in a world map.')

    parser.add_argument('-i', '--iplist', default='data/ips.lst',
                        help="File of IPs to plot, one per line")
    parser.add_argument('-g', '--geodb', default='data/geoip_ipv4.csv',
                        help="CSV document with columns: 'network,lat,lon'")
    parser.add_argument('-f', '--imagefile', default='data/worldmap_raw.jpg',
                        help="Path for the template image to plot over")
    parser.add_argument('-W', '--width', default='2058',
                        help="Width in pixels of the image file")
    parser.add_argument('-H', '--height', default='1746',
                        help="Height in pixels of the image file")
    parser.add_argument('-l', '--left', default='-180',
                        help="Degrees most to the left in the image")
    parser.add_argument('-r', '--right', default='180',
                        help="Degrees most to the right in the image")
    parser.add_argument('-b', '--bottom', default='-82',
                        help="Degrees most to the bottom in the image")
    parser.add_argument('-t', '--threading', action='store_true',
                        help="Sets the method to threading (use it with \
                              the '-n' option")
    parser.add_argument('-m', '--multiprocessing', action='store_true',
                        help="Sets the method to multiprocessing (use it with \
                              the '-w' option")
    parser.add_argument('-n', '--nsplits', default='10',
                        help="Parts the file 'geodb' is divided (threading)")
    parser.add_argument('-w', '--nworkers', default='4',
                        help="Number of workers (multiprocessing)")
    parser.add_argument('-R', '--raw', action='store_true',
                        help="Do not use threading nor multiprocessing")
    args = parser.parse_args()

    if (args.threading and args.multiprocessing) or \
       (args.threading and args.raw) or (args.multiprocessing and args.raw):
        print("Error: conflicting options.  Use '-h' or '--help' for help.",
              file=sys.stderr)
        exit(1)

    if args.raw:
        arg_method = 'raw'
    else:
        arg_method = 'threading' if args.threading else 'multiprocessing'

    arg_nsplits = int(args.nsplits) if not args.raw else 0
    arg_nworkers = int(args.nworkers) if not args.raw else 0

    # Image object instantiation
    img = GeoImage(args.imagefile,
                   width=int(args.width),
                   height=int(args.height),
                   w_deg=float(args.left),
                   e_deg=float(args.right),
                   s_deg=float(args.bottom))

    # Geo object instantiation
    geo = GeoIpMap(args.iplist, args.geodb, img,
                   method=arg_method,
                   num_workers=arg_nworkers,
                   num_splits=arg_nsplits,
                   verbose=True)

    # Do the math!
    geo.ips_to_pixels()
    geo.plot()
